mapstructure_test.go

```diff
@@ -2482,6 +2484,85 @@ func TestDecoder_MatchName(t *testing.T) {
 	}
 }
 
+func TestTypedNilPostHooks(t *testing.T) {
+	type customType1 int
+	type customType2 float64
+	type configType struct {
+		C *customType1 `mapstructure:"c"`
+		A *customType2 `mapstructure:"a"`
+	}
+
+	for i, marshalledConfig := range []map[string]interface{}{
+		{
+			"c": (*customType1)(nil),
+			"a": (*customType2)(floatPtr(42.42)),
+		},
+		{
+			"c": "",
+			"a": "42.42",
+		},
+	} {
+		t.Run(fmt.Sprintf("%v", i), func(t *testing.T) {
+			actual := &configType{}
+			decoder, err := NewDecoder(&DecoderConfig{
+				Result: actual,
+				DecodeHook: ComposeDecodeHookFunc(
+					func(f reflect.Type, t reflect.Type, data interface{}) (interface{}, error) {
+						var enum customType1
+						if f.Kind() != reflect.String || t != reflect.TypeOf(&enum) {
+							return data, nil
+						}
+						s := data.(string)
+						if s == "" {
+							// Returning an untyped nil here would cause a panic, as `from.Type()`
+							// is invalid for nil.
+							return (*customType1)(nil), nil
```

Comment:
The `panic` trace is:
```
--- FAIL: TestTypedNilPostHooks (0.00s)
    --- FAIL: TestTypedNilPostHooks/1 (0.00s)
panic: reflect: call of reflect.Value.Type on zero Value [recovered]
	panic: reflect: call of reflect.Value.Type on zero Value

goroutine 8 [running]:
testing.tRunner.func1.2({0x56fdc0, 0xc00000e120})
	/usr/local/go/src/testing/testing.go:1389 +0x24e
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1392 +0x39f
panic({0x56fdc0, 0xc00000e120})
	/usr/local/go/src/runtime/panic.go:838 +0x207
reflect.Value.Type({0x0?, 0x0?, 0x5d4e28?})
	/usr/local/go/src/reflect/value.go:2451 +0x12e
github.com/mitchellh/mapstructure.DecodeHookExec({0x571d40?, 0x5a4ed0?}, {0x0?, 0x0?, 0x570840?}, {0x564e80?, 0xc000060df0?, 0x573d00?})
	/home/adriansmares/projects/mapstructure/decode_hooks.go:47 +0x11d
github.com/mitchellh/mapstructure.ComposeDecodeHookFunc.func1({0x56a6a0?, 0x5d3ad0?, 0xc00013d358?}, {0x564e80?, 0xc000060df0?, 0xffffffffffffffff?})
	/home/adriansmares/projects/mapstructure/decode_hooks.go:69 +0x10c
github.com/mitchellh/mapstructure.DecodeHookExec({0x570840?, 0xc00006a140?}, {0x56a6a0?, 0x5d3ad0?, 0xc00013d3b8?}, {0x564e80?, 0xc000060df0?, 0x55fca9?})
	/home/adriansmares/projects/mapstructure/decode_hooks.go:51 +0xd3
github.com/mitchellh/mapstructure.(*Decoder).decode(0xc000010058, {0x55fca9, 0x1}, {0x56a6a0?, 0x5d3ad0?}, {0x564e80?, 0xc000060df0?, 0x10?})
	/home/adriansmares/projects/mapstructure/mapstructure.go:440 +0x179
github.com/mitchellh/mapstructure.(*Decoder).decodeStructFromMap(0xc000010058, {0x0, 0x0}, {0x571500?, 0xc00007e870?, 0x1?}, {0x57b740?, 0xc000060df0?, 0xc00006a120?})
	/home/adriansmares/projects/mapstructure/mapstructure.go:1382 +0x1470
github.com/mitchellh/mapstructure.(*Decoder).decodeStruct(0x570840?, {0x0, 0x0}, {0x571500?, 0xc00007e870?}, {0x57b740?, 0xc000060df0?, 0x203000?})
	/home/adriansmares/projects/mapstructure/mapstructure.go:1208 +0x4ab
github.com/mitchellh/mapstructure.(*Decoder).decode(0xc000010058, {0x0, 0x0}, {0x571500?, 0xc00007e870?}, {0x57b740?, 0xc000060df0?, 0x48?})
	/home/adriansmares/projects/mapstructure/mapstructure.go:463 +0x3cf
github.com/mitchellh/mapstructure.(*Decoder).Decode(0xc000010058, {0x571500, 0xc00007e870})
	/home/adriansmares/projects/mapstructure/mapstructure.go:398 +0xd8
github.com/mitchellh/mapstructure.TestTypedNilPostHooks.func1(0xc000115380)
	/home/adriansmares/projects/mapstructure/mapstructure_test.go:2551 +0x14e
testing.tRunner(0xc000115380, 0xc000060de0)
	/usr/local/go/src/testing/testing.go:1439 +0x102
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1486 +0x35f
FAIL	github.com/mitchellh/mapstructure	0.005s
FAIL
```

It seems to me that returning `nil` from a type decoding hook is not supported (was it supported in the first place ?)